File: src/test/java/fr/marstech/mtlinkspray/service/ShortenerServiceTest.java
package fr.marstech.mtlinkspray.service;

import fr.marstech.mtlinkspray.entity.HistoryItem;
import fr.marstech.mtlinkspray.entity.LinkItem;
import fr.marstech.mtlinkspray.entity.LinkItemTarget;
import fr.marstech.mtlinkspray.repository.LinkItemRepository;
import jakarta.servlet.http.HttpServletRequest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.testcontainers.service.connection.ServiceConnection;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.testcontainers.containers.MongoDBContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;
import org.testcontainers.utility.DockerImageName;

import java.time.LocalDateTime;
import java.util.*;

import static fr.marstech.mtlinkspray.config.TestConfig.MONGO_DB_DOCKER_IMAGE_NAME;
import static fr.marstech.mtlinkspray.utils.NetworkUtils.isValidUrl;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.when;

@Testcontainers
@SpringBootTest
class ShortenerServiceTest {

  @Container @ServiceConnection
  static MongoDBContainer mongoDBContainer =
      new MongoDBContainer(DockerImageName.parse(MONGO_DB_DOCKER_IMAGE_NAME)).withReuse(true);

  @Autowired LinkItemRepository linkItemRepository;
  @Autowired ShortenerService shortenerService;
  @MockitoBean HttpServletRequest httpServletRequest;

  @BeforeEach
  void setUp() {
    when(httpServletRequest.getHeaderNames())
        .thenReturn(
            new Enumeration<>() {
              @Override
              public boolean hasMoreElements() {
                return false;
              }

              @Override
              public String nextElement() {
                return "";
              }
            });
    when(httpServletRequest.getServerName()).thenReturn("localhost");
    when(httpServletRequest.getServerPort()).thenReturn(8080);
    when(httpServletRequest.getScheme()).thenReturn("http");
  }

  @Test
  void linkItemTest() {
    String id = UUID.randomUUID().toString();
    LinkItem linkItem =
        linkItemRepository.save(
            new LinkItem(
                id,
                LocalDateTime.now(),
                null,
                true,
                null,
                Map.of(),
                new HistoryItem(),
                List.of(),
                new LinkItemTarget("https://www.example.com")));
    Optional<LinkItem> byId = linkItemRepository.findById(id);
    //noinspection OptionalGetWithoutIsPresent
    assertEquals(linkItem.getId(), byId.get().getId());
  }

  @Test
  void shortenValidUrl() {
    String validUrl = "https://www.example.com";
    String result = shortenerService.shorten(validUrl, httpServletRequest);
    assertTrue(isValidUrl(result));
  }

  @Test
  void shortenEmptyUrl() {
    Exception exception =
        assertThrows(
            IllegalArgumentException.class,
            () -> shortenerService.shorten(null, httpServletRequest));
    assertInstanceOf(IllegalArgumentException.class, exception);
    assertEquals("URL must not be null or empty", exception.getMessage());
  }

  @Test
  void shortenInvalidUrl() {
    Exception exception =
        assertThrows(
            IllegalArgumentException.class,
            () -> shortenerService.shorten("invalid-url", httpServletRequest));
    assertInstanceOf(IllegalArgumentException.class, exception);
    assertEquals("Invalid URL: invalid-url", exception.getMessage());
  }
}

File: src/test/kotlin/fr/marstech/mtlinkspray/service/ShortenerServiceTest.kt
package fr.marstech.mtlinkspray.service

import fr.marstech.mtlinkspray.entity.HistoryItem
import fr.marstech.mtlinkspray.entity.LinkItem
import fr.marstech.mtlinkspray.entity.LinkItemTarget
import fr.marstech.mtlinkspray.repository.LinkItemRepository
import fr.marstech.mtlinkspray.utils.NetworkUtils.isValidUrl
import jakarta.servlet.http.HttpServletRequest
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.springframework.data.crossstore.ChangeSetPersister
import java.time.LocalDateTime
import java.util.*

@ExtendWith(MockitoExtension::class)
internal class ShortenerServiceTest {
    private val linkItemRepository = mock<LinkItemRepository>()
    private val httpServletRequest = mock<HttpServletRequest>()
    private val randomIdGeneratorService = mock<RandomIdGeneratorService>()
    private val shortenerService = ShortenerServiceImpl(linkItemRepository, randomIdGeneratorService)

    @BeforeEach
    fun setUp() {
        whenever(httpServletRequest.headerNames)
            .thenReturn(object : Enumeration<String?> {
                override fun hasMoreElements(): Boolean = false
                override fun nextElement(): String = ""
            })
        whenever(httpServletRequest.serverName).thenReturn("localhost")
        whenever(httpServletRequest.serverPort).thenReturn(8080)
        whenever(httpServletRequest.scheme).thenReturn("http")
        whenever(randomIdGeneratorService.getGeneratedFreeId()).thenReturn("test-id")
    }

    @Test
    fun linkItemTest() {
        val id = UUID.randomUUID().toString()
        val linkItem = LinkItem(
            id,
            LocalDateTime.now(),
            null,
            true,
            null,
            mutableMapOf(),
            HistoryItem(),
            mutableListOf(),
            LinkItemTarget("https://www.example.com")
        )
        whenever(linkItemRepository.save(any())).thenReturn(linkItem)
        whenever(linkItemRepository.findById(id)).thenReturn(Optional.of(linkItem))
        linkItemRepository.save(linkItem)
        val found = linkItemRepository.findById(id)
        Assertions.assertNotNull(found)
        Assertions.assertTrue(found.isPresent)
        Assertions.assertEquals(id, found.get().id)
    }

    @Test
    fun shortenValidUrl() {
        val id = "uid"
        val linkItem = LinkItem(
            id,
            LocalDateTime.now(),
            null,
            true,
            null,
            mutableMapOf(),
            HistoryItem(),
            mutableListOf(),
            LinkItemTarget("https://www.example.com")
        )
        whenever(linkItemRepository.save(any())).thenReturn(linkItem)
        whenever(linkItemRepository.findById(id)).thenReturn(Optional.of(linkItem))
        whenever(randomIdGeneratorService.getGeneratedFreeId()).thenReturn(id)
        shortenerService.shorten("https://www.example.com", httpServletRequest).let {
            Assertions.assertNotNull(it)
            Assertions.assertTrue(it.contains("http://localhost:8080/"))
            Assertions.assertTrue(isValidUrl(it))
        }
    }

    @Test
    fun shortenInvalidUrl() {
        whenever(randomIdGeneratorService.getGeneratedFreeId()).thenReturn("invalid-id")
        whenever(linkItemRepository.save(any())).thenReturn(null)
        val exception: Exception? = Assertions.assertThrows(
            NullPointerException::class.java
        ) { shortenerService.shorten("invalid-url", httpServletRequest) }
        Assertions.assertInstanceOf(NullPointerException::class.java, exception)
    }

    @Test
    fun getTargetValidUid() {
        val id = "uid"
        val url = "https://www.example.com"
        val linkItem = LinkItem(
            id,
            LocalDateTime.now(),
            null,
            true,
            null,
            mutableMapOf(),
            HistoryItem(),
            mutableListOf(),
            LinkItemTarget(url)
        )
        whenever(linkItemRepository.save(any())).thenReturn(linkItem)
        whenever(linkItemRepository.findById(id)).thenReturn(Optional.of(linkItem))
        whenever(randomIdGeneratorService.getGeneratedFreeId()).thenReturn(id)
        val shortened = shortenerService.shorten(url, httpServletRequest)
        val uid = shortened.substring(shortened.lastIndexOf('/') + 1)
        val target = shortenerService.getTarget(uid)
        Assertions.assertEquals(url, target)
    }

    @Test
    fun getTargetInvalidUid_throwsException() {
        whenever(linkItemRepository.findById(any())).thenReturn(Optional.empty())
        Assertions.assertThrows(
            ChangeSetPersister.NotFoundException::class.java
        ) { shortenerService.getTarget("nonexistent-uid") }
    }

    @Test
    fun shortenAndRetrievePersistedLink() {
        val id = "uid"
        val url = "https://persisted.com"
        val linkItem = LinkItem(
            id,
            LocalDateTime.now(),
            null,
            true,
            null,
            mutableMapOf(),
            HistoryItem(),
            mutableListOf(),
            LinkItemTarget(url)
        )
        whenever(linkItemRepository.save(any())).thenReturn(linkItem)
        whenever(linkItemRepository.findById(id)).thenReturn(Optional.of(linkItem))
        whenever(randomIdGeneratorService.getGeneratedFreeId()).thenReturn(id)
        val shortened = shortenerService.shorten(url, httpServletRequest)
        val uid = shortened.substring(shortened.lastIndexOf('/') + 1)
        val item = linkItemRepository.findById(uid)
        Assertions.assertTrue(item.isPresent)
        Assertions.assertEquals(url, item.get().target.targetUrl)
    }
}
